src/org/traccar/api/BaseResource.java

```diff
@@ -29,4 +37,35 @@ protected long getUserId() {
         }
         return 0;
     }
+
+    protected void handlePermission(Map<String, Long> entity, boolean link) throws SQLException {
+        if (entity.size() != 2) {
+            throw new IllegalArgumentException();
+        }
+        Iterator<String> iterator = entity.keySet().iterator();
+        String owner = iterator.next();
+        String property = iterator.next();
+
+        long ownerId = entity.get(owner);
+        long propertyId = entity.get(property);
+
+        if (!link && DataManager.makeName(owner).equals(Context.TYPE_USER)
+                && DataManager.makeName(property).equals(Context.TYPE_DEVICE)) {
+            if (getUserId() != ownerId) {
+                Context.getPermissionsManager().checkUser(getUserId(), ownerId);
+            } else {
+                Context.getPermissionsManager().checkAdmin(getUserId());
+            }
+        } else {
+            Context.getPermissionsManager().checkPermission(owner, getUserId(), ownerId);
+        }
+        Context.getPermissionsManager().checkPermission(property, getUserId(), propertyId);
+
+        Context.getDataManager().linkObject(owner, ownerId, property, propertyId, link);
+    }
+
+    protected void linkNew(BaseModel entity) throws SQLException {
```

Comment:
I would recommend to rename this method to `linkNewEntity` to make it more clear in places where we use it.